### PR TITLE
nsd plugin

### PR DESCRIPTION
**Pull request by 383c57**
**nsd**
https://www.nlnetlabs.nl/projects/nsd/

Tested on nsd-4.1.14(from EPEL), CentOS7

Thank you.

### CODE REVIEW

**Comment by l2isbad** [comment]
Yo @383c57 

1. dict comprehension doesn't work in python 2.6
2. I don't think you need `try except`. The only exception i see if `_get_raw_data` returns `None` and regex findall will raise exception

a)
```python
lines = self._get_raw_data() or list()
```
b) better option
```python
lines = self._get_raw_data()
if not lines:
    return None
```

And can you post here output for `time nsd-control stats_noreset`

**Comment by l2isbad** [comment]
@ktsaou 
your words
> If you do it, replace the . with _ at the dimension id and give it with a dot as the name.

Is it really necessary? If so @383c57 you need to replace all dots with _

**Comment by ktsaou** [comment]
> > If you do it, replace the . with _ at the dimension id and give it with a dot as the name.
> Is it really necessary? If so @383c57 you need to replace all dots with _

Well, in various backends, a metric is composed with something like "PREFIX.HOST.CHART_TYPE.CHART_ID.DIMENSION_ID". Using dimension IDs with dots, may confuse a lot of guys out there, that may expect components to be `.` separated.

I think this is the only reason...

**Comment by 383c57** [comment]
Hi.
Thank you for your comments.

@l2isbad 
```
# time nsd-control stats_noreset
server0.queries=62
server1.queries=128
num.queries=190
time.boot=936863.619160
time.elapsed=936863.619160
size.db.disk=25856
size.db.mem=15664
size.xfrd.mem=20995504
size.config.disk=234
size.config.mem=5232
num.type.A=161
num.type.NS=0
num.type.MD=0
num.type.MF=0
num.type.CNAME=0
num.type.SOA=6
num.type.MB=0
num.type.MG=0
num.type.MR=0
num.type.NULL=0
num.type.WKS=0
num.type.PTR=0
num.type.HINFO=1
num.type.MINFO=0
num.type.MX=6
num.type.TXT=3
num.type.RP=0
num.type.AFSDB=0
num.type.X25=0
num.type.ISDN=0
num.type.RT=0
num.type.NSAP=0
num.type.SIG=0
num.type.KEY=0
num.type.PX=0
num.type.AAAA=1
num.type.LOC=0
num.type.NXT=0
num.type.SRV=5
num.type.NAPTR=0
num.type.KX=0
num.type.CERT=0
num.type.DNAME=0
num.type.OPT=0
num.type.APL=0
num.type.DS=0
num.type.SSHFP=0
num.type.IPSECKEY=0
num.type.RRSIG=0
num.type.NSEC=0
num.type.DNSKEY=0
num.type.DHCID=0
num.type.NSEC3=0
num.type.NSEC3PARAM=0
num.type.TLSA=0
num.type.CDS=0
num.type.CDNSKEY=0
num.type.OPENPGPKEY=0
num.type.CSYNC=0
num.type.SPF=0
num.type.NID=0
num.type.L32=0
num.type.L64=0
num.type.LP=0
num.type.EUI48=0
num.type.EUI64=0
num.type.TYPE252=5
num.type.TYPE255=2
num.opcode.QUERY=190
num.class.IN=190
num.rcode.NOERROR=173
num.rcode.FORMERR=0
num.rcode.SERVFAIL=0
num.rcode.NXDOMAIN=0
num.rcode.NOTIMP=0
num.rcode.REFUSED=12
num.rcode.YXDOMAIN=0
num.edns=189
num.ednserr=0
num.udp=18
num.udp6=167
num.tcp=0
num.tcp6=5
num.answer_wo_aa=0
num.rxerr=0
num.txerr=0
num.raxfr=0
num.truncated=0
num.dropped=0
zone.master=4
zone.slave=0

real    0m0.086s
user    0m0.019s
sys     0m0.011s
```

**Comment by l2isbad** [comment]
thks @383c57 !
Are there any chance for regex to fail? If yes we should return None instead of dict with 3 keys (setdefault).

**Comment by 383c57** [comment]
There are probably no.

**Comment by l2isbad** [comment]
@383c57 

> usermod -aG nsd netdata

We can add netdata to group during installation.

You can do it in [netdata-installer.sh](https://github.com/firehol/netdata/blob/master/netdata-installer.sh) (we did it for` adm, haproxy, varnish, nginx, docker` groups)

```bash
grep -i varnish netdata-installer.sh 
NETDATA_ADDED_TO_VARNISH=0
    portable_add_user_to_group varnish  netdata && NETDATA_ADDED_TO_VARNISH=1
getent group varnish > /dev/null
if [ $? -eq 0 -a "${NETDATA_ADDED_TO_VARNISH}" = "1" ]
    echo "You may also want to remove the netdata user from the varnish group"
    echo "   gpasswd -d netdata varnish"
```

**Comment by l2isbad** [comment]
And i did some fixes for [ExecutableService](https://github.com/firehol/netdata/pull/1905).
Can you check your plugin still works fine with new version?

File: python.d/nsd.chart.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        lines = self._get_raw_data()
+        if not lines:
+            return None
+
+        r = self.regex
+        stats = dict((k.replace('.', '_'), int(v))
+                     for k, v in r.findall(''.join(lines)))
+        stats.setdefault('num.opcode.NOTIFY', 0)
+        stats.setdefault('num.type.TYPE252', 0)
+        stats.setdefault('num.type.TYPE255', 0)
```
Comment (ktsaou) [inline python.d/nsd.chart.py]: dots or underscores?

Comment (383c57) [inline python.d/nsd.chart.py]: Thank you for your review.
sorry. replace dot with _ failed.

**Comment by 383c57** [comment]
Thank you @l2isbad 
> And i did some fixes for ExecutableService.
> Can you check does your plugin still works fine with new version?

nsd plugin works fine.

**Comment by l2isbad** [comment]
gj @383c57 
And last but not least.
Please update python [README file](https://github.com/firehol/netdata/blob/master/python.d/README.md)
And to the [wiki page](https://github.com/firehol/netdata/wiki/Add-more-charts-to-netdata) (open for anyone to edit)

**Comment by ktsaou** [comment]
nice work!
When the docs are updated, please post here and I'll merge it.

**Comment by 383c57** [comment]
docs updated

**Comment by ktsaou** [comment]
merged! thank you!

**Comment by 383c57** [comment]
Thank you!

**Pull request merged**